fix: number of google pages results, limited perms, improvements from WCC

MQ37 opened the pull request:
closes https://github.com/apify/rag-web-browser/issues/80

:warning: MERGE ONLY AFTER https://github.com/apify/rag-web-browser/pull/82 IS MERGED :warning: 

Fixes the google pages number of results scraping. Instead of using the ?num param that is now ignored by the google it uses the ?start for offset pagination to gather the google result pages urls. Implementation based on https://github.com/apify-store/google-search

MQ37 committed:
fix: number of google pages results

jirispilka reviewed (commented):
if possible I would use reference implementation, I think it is better to have one source of truth

These are LLM complains:
The reference implementation is more robust because it:
- Validates pagination links exist in HTML
- Tracks page numbers to avoid non-existent pages
- Uses multiple fallback conditions
- Handles Google's inconsistent pagination behavior

MQ37 commented on src/crawlers.ts:
```diff
@@ -83,27 +83,62 @@ export async function createAndStartSearchCrawler(
+                // Queue the next page
+                const nextPage = currentPage + 1;
+                const nextOffset = nextPage * 10;
+                log.info(`Queueing next page (${nextPage + 1}/${totalPages}) with offset ${nextOffset}`);
```
if you mean the one in the log it is that the var is index and we want to print human readable number so + 1. Think of first page of a book is index 0 but from human readable point its page 1 (means 0 + 1)

MQ37 replied:
updated

matyascimbulka commented on src/utils.ts:
```diff
@@ -90,15 +93,18 @@ export function createSearchRequest(
+    collectedResults: OrganicResult[] = [],
+    currentPage = 0,
+    totalPages = Math.ceil(maxResults / 10) + 1,
 ): RequestOptions<SearchCrawlerUserData> {
```
This is quite heavy method signature. Most of these fields are just passed in to create the user data object. We can make this method accept an object with the SearchCrawlerUserData type and use that to ferry the data.

MQ37 replied:
refactored, thanks :+1:

matyascimbulka reviewed (changes requested):
I like moving the validation of the results to the collection also the move to fix pagination looks good. But I have found some duplicate code and some opportunities to make the code cleaner.

MQ37 committed:
Update src/crawlers.ts

Co-authored-by: Matyáš Cimbulka <matyas.cimbulka@apify.com>

MQ37 committed:
Update src/crawlers.ts

Co-authored-by: Matyáš Cimbulka <matyas.cimbulka@apify.com>

MQ37 committed:
refactor: user data handling for the serp pagination

MQ37 committed:
refactor: remove duplicate code

MQ37 committed:
fix: update request handler to use addRequests for search crawler

matyascimbulka commented on src/utils.ts:
```diff
@@ -75,40 +75,53 @@ export function randomId() {
+        maxResults: number;
+        contentCrawlerKey: string;
+        contentScraperSettings: ContentScraperSettings;
+    },
```
This works but its not very Typescript way to do this I'm not huge fan of this because you're declaring the some properties twice.

Here is some Typescript magic that can do the same thing without declaring the properties twice:
```typescript
type Optional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

export type CreateSearchRequestUserData = Optional<SearchCrawlerUserData, 'timeMeasures' | 'collectedResults' | 'currentPage' | 'totalPages'>;
```

matyascimbulka reviewed (changes requested):
Looks better but still found some issues that would most likely cause issues down the line. It would be good to see some test runs of the entire PR stack.

matyascimbulka commented on src/utils.ts:
```diff
@@ -75,40 +75,53 @@ export function randomId() {
-            contentScraperSettings,
-            responseId,
+            maxResults: userData.maxResults,
+            timeMeasures: userData.timeMeasures || [],
```
I just realize that I'm wrong here. It won't fail.

matyascimbulka reviewed (approved):
I just realize that my comments about it failing are wrong. Still it be nice to see some test runs.

jirispilka commented:
> Regarding the "reference" implementation Ondra mentioned that the page tracking logic is legacy and they just left it there, so our simpler implementation should be fine.

Ok, thanks for the clarification. IMHO it is always better to ask than miss some edge cases

MQ37 committed:
make code more typescripty

MQ37 committed:
add var for better readability

MQ37 committed:
feat: update libs for limited perms, fix and speed up Dockerfile build, fix wrong imports, firefox patches and policies, NY TZ (#82)

* feat: update libs for limited perms, fix and speed up Dockerfile build, fix wrong imports

* magic number to const

* lint

* fix lint, update crawlee

* add firefox policies, patches, set NY TZ

* add ghostery blocker

* Squashed commit of the following:

commit 7aadecda87a74a5a61494ca5664cec44308e7569
Author: MQ37 <themq37@gmail.com>
Date:   Wed Nov 26 11:14:09 2025 +0100

    add var for better readability

commit fd02b0a65ddf83e151d46538bf12ec88da8d95c5
Author: MQ37 <themq37@gmail.com>
Date:   Wed Nov 26 11:09:57 2025 +0100

    make code more typescripty

MQ37 committed:
fix playwright version, fix types.ts

MQ37 committed:
fix playwright version for the patch

MQ37 committed:
fix merge issue

matyascimbulka reviewed (approved):
Looks good, just found small nit.

MQ37 committed:
Update src/utils.ts

Co-authored-by: Matyáš Cimbulka <matyas.cimbulka@apify.com>
